File: backend/app/run/channels.py
"""Per-agent channel between an orchestrator and an agent.

The orchestrator message port stays the end of the run. Each agent gets
`channel:<agentId>`. A stored message edge from orchestrator to agent is
rewritten so older graphs keep running.
"""

from __future__ import annotations

from app.run.graph_models import AgentNetworkDocument, GraphEdge


def channel_handle(agent_id: str) -> str:
    return f"channel:{agent_id}"


def agent_id_from_channel(handle: str) -> str | None:
    prefix = "channel:"
    if not handle.startswith(prefix):
        return None
    agent_id = handle[len(prefix) :]
    return agent_id or None


def normalize_channel_edges(doc: AgentNetworkDocument) -> AgentNetworkDocument:
    by_id = {node.id: node for node in doc.nodes}
    changed = False
    edges: list[GraphEdge] = []
    for edge in doc.edges:
        src = by_id.get(edge.source)
        dst = by_id.get(edge.target)
        if (
            src is not None
            and dst is not None
            and src.type == "orchestrator"
            and dst.type == "agent"
            and edge.source_handle == "message"
            and edge.target_handle == "message"
        ):
            changed = True
            edges.append(
                edge.model_copy(
                    update={
                        "source_handle": channel_handle(dst.id),
                        "target_handle": "channel",
                    }
                )
            )
        else:
            edges.append(edge)
    if not changed:
        return doc
    return doc.model_copy(update={"edges": edges})

File: frontend/src/modules/network/schema/channels.ts
import type { AgentNetworkDocument, GraphNode } from '@/modules/network/model/document';

/** Free orchestrator port. A drop on it becomes `channel:<agentId>`. */
export const CHANNEL_GHOST_ID = 'channel';

export function channelHandle(agentId: string): string {
  return `channel:${agentId}`;
}

export function agentIdFromChannel(handle: string | null | undefined): string | null {
  if (!handle?.startsWith('channel:')) return null;
  const id = handle.slice('channel:'.length);
  return id || null;
}

/** Old graphs hung every agent on the orchestrator message port. That port is only the end now. */
export function normalizeChannelEdges(doc: AgentNetworkDocument): AgentNetworkDocument {
  const byId = new Map(doc.nodes.map((node) => [node.id, node]));
  let changed = false;
  const edges = doc.edges.map((edge) => {
    const source = byId.get(edge.source);
    const target = byId.get(edge.target);
    if (
      source?.type === 'orchestrator' &&
      target?.type === 'agent' &&
      edge.sourceHandle === 'message' &&
      edge.targetHandle === 'message'
    ) {
      changed = true;
      return { ...edge, sourceHandle: channelHandle(target.id), targetHandle: CHANNEL_GHOST_ID };
    }
    return edge;
  });
  if (!changed) return doc;
  return { ...doc, edges };
}

export function channelLabel(agent: GraphNode | undefined, agentId: string): string {
  const name = agent && typeof agent.data.displayName === 'string' ? agent.data.displayName.trim() : '';
  return name || agentId;
}
